### ISSUE TITLE
Adjoint gradient returns zero when output uses interp1d

### ISSUE THREAD

**Issue by martinjrobins** [martinjrobins/diffsol#319]
## Bug

When the output expression contains `interp1d`, the discrete adjoint backward pass (`solve_adjoint_bkwd`) returns zero for all parameter gradients, even though forward sensitivity (`solve_fwd_sens`) produces correct non-zero values.

## MRE

```python
import pybamm, numpy as np, pydiffsol as ds

model = pybamm.BaseModel("test")
x = pybamm.Variable("x"); y = pybamm.Variable("y")
k = pybamm.InputParameter("k")
model.rhs = {x: -k * x}
model.algebraic = {y: x - y}
model.initial_conditions = {x: pybamm.Scalar(1.0), y: pybamm.Scalar(0.7)}

# Output uses interp1d — adjoint returns zero
data_v = np.exp(-0.5 * np.linspace(0, 2, 11))
target = pybamm.DiscreteTimeData(np.linspace(0, 2, 11), data_v, "target")
model.variables = {"diff": target - y}  # interp1d involved

sim = pybamm.Simulation(model, solver=pybamm.IDAKLUSolver())
exporter = pybamm.DiffSLExport(sim)
code = exporter.to_diffeq(["diff"])

# Forward sensitivity: works fine
ode_fs = ds.Ode(code, matrix_type=ds.faer_sparse, scalar_type=ds.f64,
                linear_solver=ds.lu, ode_solver=ds.bdf)
ode_fs.rtol = 1e-6; ode_fs.atol = 1e-6
ode_fs.sens_rtol = None; ode_fs.sens_atol = None
sol_fs = ode_fs.solve_fwd_sens(np.array([0.7]), np.array([2.0]))
print(f"Fwd sens: {sol_fs.sens[0][0,0]:.8f}")  # -> non-zero (e.g. 0.49319234)

# Adjoint: returns zero
t_eval = np.linspace(0, 2, 11)
ode_adj = ds.Ode(code, matrix_type=ds.faer_sparse, scalar_type=ds.f64,
                 linear_solver=ds.lu, ode_solver=ds.bdf)
ode_adj.rtol = 1e-6; ode_adj.atol = 1e-6
ode_adj.sens_rtol = None; ode_adj.sens_atol = None
ode_adj.param_rtol = 1e-4; ode_adj.param_atol = 1e-6
sol_adj, ckpt = ode_adj.solve_adjoint_fwd(np.array([0.7]), t_eval)
grad = ode_adj.solve_adjoint_bkwd(sol_adj, ckpt, np.ones((1, len(t_eval))))
print(f"Adjoint: {np.asarray(grad).ravel()}")  # -> [0.]
```

Removing `interp1d` from the output (e.g. output = `y` or `pow(y, 2)`) makes adjoint work correctly.

## Root cause

The `interp1d` function in the DiffSL compiler likely does not implement `NonLinearOpAdjoint` or `jac_transpose_mul_inplace` correctly. When the adjoint backward pass calls `integrate_delta_g` at each measurement time, `out.jac_transpose_mul_inplace(state, t, dgdu, result)` should propagate the output gradient to the adjoint state λ. But when `out` involves `interp1d`, this call produces zero, so λ never gets the discrete jump and the entire backward pass computes zero gradients.

## Affected use case

This breaks adjoint sensitivity for all PyBaMM models whose loss function references `DiscreteTimeData` (interpolated target data), which is the standard pattern for parameter estimation — e.g. `loss = sum((V_data(t) - V_model(t))^2)`.

Forward sensitivity is unaffected and works correctly for the same models.
